nutils/function.py

```diff
@@ -3276,6 +3276,136 @@ def _diagonalize(self, axis, newaxis):
+
+  >>> class IntPower(Custom):
+  ...
+  ...   def __init__(self, base, exponent):
+  ...     # We are being nit-picky here and cast `exponent` to an `int` without
+  ...     # truncation.
+  ...     exponent = exponent.__index__()
+  ...     super().__init__(args=(base, exponent), shape=base.shape, dtype=float)
+  ...
+  ...   @staticmethod
+  ...   def evalf(base, exponent):
+  ...     # `base` is evaluated to a `numpy.ndarray` because we passed `base`
+  ...     # as an `Array` to the constructor. `exponent`, however, is untouched
+  ...     # because it is not an `Array`.
+  ...     return base**exponent
+  ...
+  ...   @staticmethod
+  ...   def partial_derivative(iarg, base, exponent):
+  ...     if iarg == 0:
+  ...       if exponent == 0:
+  ...         # The return value is cast to `Array` by the caller.
+  ...         return 0
+  ...       else:
+  ...         pd = exponent * IntPower(base, exponent - 1)
+  ...         return functools.reduce(diagonalize, range(base.ndim), pd)
+  ...     else:
+  ...       # We don't implement the derivative to `exponent`, because this is
+  ...       # a constant `int`.
+  ...       raise NotImplementedError
+  ...
+  >>> IntPower(asarray([2., 3.]), 2).eval()
+  array([[ 4.,  9.]])
+  >>> derivative(IntPower(a, 2), a).eval(a=numpy.array([2., 3.]))
+  array([[[ 4.,  0.],
+          [ 0.,  6.]]])
+  '''
+
+  def __init__(self, args, shape, dtype):
+    self._args = tuple(args)
+    evaluable_args = tuple(arg if isinstance(arg, Evaluable) else EvaluableConstant(arg) for arg in self._args)
+    super().__init__(args=evaluable_args, shape=shape, dtype=dtype)
+
+  def partial_derivative(self, iarg, *args):
+    '''Return the partial derivative of this function to :class:`Custom` constructor argument number ``iarg``.
+
+    This method is only called for those arguments that are instances of
+    :class:`Array` with dtype :class:`float` and have the derivative target as
+    a dependency. It is therefor allowed to omit an implementation for some or
+    all arguments if the above conditions are not met.
+
+    Parameters
+    ----------
+    iarg : :class:`int`
+        The index of the argument to compute the derivative for.
+    *args
+        The arguments as passed to the constructor of :class:`Custom`.
+
+    Returns
+    -------
+    :class:`Array` or similar
+        The partial derivative of this function to the given argument.
+    '''
+
+    raise NotImplementedError('The partial derivative of {} to argument {} (counting from 0) is not defined.'.format(type(self).__name__, iarg)) # pragma: nocover
+
+  def _derivative(self, var, seen):
+    if self.dtype != float:
+      return super()._derivative(var, seen)
+    result = Zeros(self.shape + var.shape, dtype=self.dtype)
+    for iarg, arg in enumerate(self._args):
+      if not isinstance(arg, Array) or arg.dtype != float or var not in arg.dependencies and var != arg:
+        continue
+      d1 = asarray(self.partial_derivative(iarg, *self._args))
```

Comment:
Let's add a shape check here to avoid obscure errors:
```
if d1.shape != self.shape + arg.shape:
  raise Exception('invalid partial derivative to argument #{}: expected shape {}, got {}'.format(iarg, self.shape + arg.shape, d1.shape))
```

Reply:
Tentatively added as e948305b.

Reply:
Accepted with slight change in error message.